### PR TITLE
perf(prometheus) optimize proxy path long tail latency when scraping

### PR DESCRIPTION
**Pull request by fffonion**
### Summary

Performance optimizations

### Full changelog

* perf(prometheus) optimize performance of export functions
* perf(prometheus) store most of gauges to local tables and skip shdict
* perf(prometheus) turn off high cardinality metrics by default

### Issue reference

FTI-4083 FT-2956

### CODE REVIEW

File: kong/plugins/prometheus/prometheus.lua [hunk 1]
```diff
@@ -720,6 +766,11 @@ local function register(self, name, help, label_names, buckets, typ)
     return
   end
 
+  if typ ~= TYPE_GAUGE and local_storage then
+    ngx.log(ngx.ERR, "Cannot use local_storage metrics for non Gauge type")
```
Comment (chronolaw) [inline kong/plugins/prometheus/prometheus.lua:770]: should be `ngx_log`

File: kong/plugins/prometheus/prometheus.lua [hunk 2]
```diff
@@ -692,7 +738,7 @@ end
 --
 -- Returns:
 --   a new metric object.
-local function register(self, name, help, label_names, buckets, typ)
+local function register(self, name, help, label_names, buckets, typ, local_storage)
   if not self.initialized then
     ngx.log(ngx.ERR, "Prometheus module has not been initialized")
```
Comment (chronolaw) [inline kong/plugins/prometheus/prometheus.lua:743]: localize `ngx.log` `ngx.ERR`

File: kong/plugins/prometheus/prometheus.lua [hunk 3]
```diff
@@ -790,38 +861,68 @@ function Prometheus:metric_data()
   -- Force a manual sync of counter local state (mostly to make tests work).
   self._counter:sync()
 
-  local keys = self.dict:get_keys(0)
+  local keys = {} -- self.dict:get_keys(0)
+  local count = #keys
+  for k, v in pairs(self.local_metrics) do
```
Comment (chronolaw) [inline kong/plugins/prometheus/prometheus.lua:866]: localize `pairs`

File: kong/plugins/prometheus/prometheus.lua [hunk 4]
```diff
@@ -790,38 +861,68 @@ function Prometheus:metric_data()
...
+    keys[count+1] = k
+    count = count + 1
+  end
   -- Prometheus server expects buckets of a histogram to appear in increasing
   -- numerical order of their label values.
   table.sort(keys)
```
Comment (chronolaw) [inline kong/plugins/prometheus/prometheus.lua:872]: localzie `table.sort`

File: kong/plugins/prometheus/prometheus.lua [hunk 5]
```diff
@@ -790,38 +861,68 @@ function Prometheus:metric_data()
...
+    end
+
+    if output_count >= 100 or not data then
+        ngx_print(output)
+        output_count = 0
+        tb_clear(output)
+    end
+  end
+
   for _, key in ipairs(keys) do
```
Comment (chronolaw) [inline kong/plugins/prometheus/prometheus.lua:893]: localize `ipairs`

**Comment by sichvoge** [comment]
@fffonion I guess this also needs to be documented inside the Prometheus Plugin page right? cc/ @ryanmoore @hfukada

**Comment by fffonion** [comment]
@sichvoge Definitely, I will start docs once this PR is finalized.

**Comment by fffonion** [comment]
i will address reviews next week, and add some Targets in performance test to validate local stored gauge's prerformance

File: kong/plugins/prometheus/prometheus.lua [hunk 6]
```diff
@@ -761,16 +834,34 @@ local function register(self, name, help, label_names, buckets, typ)
   return metric
 end
 
+-- inspired by https://github.com/Kong/kong/blob/2.8.1/kong/tools/utils.lua#L1430-L1446
+-- limit to work only in rewrite, access, content and timer
+local yield
```
Comment (mayocream) [inline kong/plugins/prometheus/prometheus.lua:839]: can we reuse the one in `kong.tools.utils`?

**Review by mayocream — APPROVED** [review]
Looks OK to me

**Review by mayocream — APPROVED** [review]

**Comment by flrgh** [comment]
@fffonion sorry to drag up such an old PR, but I wanted to check: can #8844 be closed out due to this PR, or is there still more work to be performed before that issue is ready to close?

**Pull request merged**